### ISSUE TITLE
Docker-compose containers can't resolve container names

### ISSUE THREAD

**Issue by lawndoc** [actions/runner-images#6026]
### Description

When deploying multiple containers with docker that are supposed to communicate with each other, referencing their hostname by the container's name does not resolve. The docker daemon should make container hostnames available to all the containers in the runtime.

I've made a docker-compose.test.yml file in [the repo I am having issues with](https://github.com/commander-framework/commander), and running the tests works locally on my machine, but the same compose file fails in GH Actions due to container name resolution not working. Below is a sample failure message:

```
>           raise ConnectionError(e, request=request)
E           requests.exceptions.ConnectionError: HTTPSConnectionPool(host='nginx', port=443): Max retries exceeded with url: /admin/login (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fbd038cb6a0>: Failed to establish a new connection: [Errno -***] Name or service not known'))
```

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions

### Virtual environments affected

- [ ] Ubuntu 18.04
- [X] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] macOS 12
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Image version and build link

[Failed run link](https://github.com/commander-framework/commander/runs/7694294427?check_suite_focus=true)

Version number: 20220729.1

### Is it regression?

Not that I know of

### Expected behavior

Containers should be able to resolve each other's container names as IPs on the docker daemon's internal network.

### Actual behavior

Containers cannot resolve each other's addresses when referenced by container name (ex. 'nginx').

### Repro steps

To reproduce the issue:
1. Create a docker-compose file with a couple of containers
2. Deploy the containers using `docker-compose`
3. In one of the containers run a curl request or python script that is targeting the service in the other container (using the container's name as the hostname).

**Comment by igorboskovic3** [comment]
Hi @lawndoc we will take a look, thank you for reporting this.

**Comment by al-cheb** [comment]
Hello @lawndoc.
I am not able reproduce the issue - https://github.com/al-cheb/docker-example:

```
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - run: |
          cd exmp && docker compose -p app up --detach
      - run: |
          echo "curl http://server:1234"
          docker exec app-client-1 curl -s http://server:1234
          echo "curl http://app-server-1:1234"
          docker exec app-client-1 curl -s http://app-server-1:1234
          echo "host server:"
          docker exec app-client-1 host server
          echo "host app-server-1:"
          docker exec app-client-1 host app-server-1
      - run: |
          docker logs app-client-1
```

![image](https://user-images.githubusercontent.com/47745270/183594986-ff95fbe6-6317-48ca-a568-841ae1d4c60c.png)

**Comment by lawndoc** [comment]
Interesting. Maybe it makes a difference if containers are tied to multiple networks? Sample from my compose file:
```yml
services:
  ...
  sut:
    networks:
      - frontend
  ...
  nginx:
    networks:
      - frontend
      - application
  ...
networks:
  frontend:
  application:
  backend:
```

**Comment by al-cheb** [comment]
@lawndoc,
I have updated my repo to add multiple networks https://github.com/al-cheb/docker-example/blob/main/docker-compose.yml: **client ->[net:frontend] nginx[port:80] ->[net:backend] backend[port:1234]**

```
      - run: |
          echo "[client] curl http://server"
          docker exec app-client-1 curl -vs http://server
          echo "[server] curl -v http://backend:1234"
          docker exec app-server-1 curl -vs http://backend:1234
```

![image](https://user-images.githubusercontent.com/47745270/183689775-4083c212-b2a4-4f99-af03-f6ba3a9342b9.png)

**Comment by lawndoc** [comment]
There must be a misconfiguration or something in my workflow or GitHub repo then... Sorry for the wasted time.

**Comment by nkleinbaer** [comment]
@lawndoc I'm running into a similar issue, were you ever able to figure out the cause of this and/or a solution?

**Comment by lawndoc** [comment]
> @lawndoc I'm running into a similar issue, were you ever able to figure out the cause of this and/or a solution?

In my case, it was actually caused by a separate issue. The container I was trying to contact was failing to start when running in GitHub Actions, so it wasn't being resolved by the Docker daemon. I'd recommend creating an artifact of the logs of your containers, and after the failed run check and make sure that all of the containers are starting properly.

You could also assign a static IP to each container (something like [this](https://gist.github.com/natcl/3d881d00a56c8a961e6dab8ba51a5a37)) and try to reference that IP instead and see if it still fails.

**Comment by lawndoc** [comment]
It was a subtle issue because the container wasn't failing to start when testing locally. In my case, I had a missing secret in the environment I was testing in.

**Comment by nkleinbaer** [comment]
Thanks for the reply, @lawndoc! 

In my case I was able to confirm the containers were successfully starting, and at least based on `docker network ls` they appeared to be attached to the appropriate bridge networks. I think the issue may have been that I was running the entire job inside a container [as described here](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container). I'm guessing there may have been some issue stemming from running "docker inside of docker", because once I switched to running the job directly on the runner the issue went away. I'm not certain however, as when I made this change I also switched to using `docker compose` instead of `docker-compose`, so that is a confounding variable.
